### PR TITLE
Sarvesh/merkel patricia proof generator

### PR DESCRIPTION
**Pull request by 0xsarvesh**
1. Exposed service to generate Account Proof
2. Exposed service to generate Storage Proof
3. Generating storage proof in batch form
4. Using singleton instance of leveldb
5. Using response helper and logger from openst-base
6. Unit test cases Proof generation, db factory, helper, library etc.  

#140

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,6 +39,10 @@
     "continuation-local-storage": "3.2.1",
     "debug": "2.2.0",
     "edit-json-file": "1.0.8",
+    "ethereumjs-util": "^5.2.0",
+    "leveldown": "^4.0.1",
+    "levelup": "^3.0.1",
+    "merkle-patricia-tree": "^2.3.1",
     "node-cmd": "3.0.0",
```
Comment (abhayks1) [inline package.json]: Let's freeze these npm versions.

File: package.json [hunk 2]
```diff
@@ -49,6 +55,8 @@
     "chai": "4.1.2",
     "ink-docstrap": "1.3.2",
     "jsdoc": "3.5.5",
-    "mocha": "5.0.0"
+    "mocha": "5.0.0",
+    "mock-require": "^3.0.2",
+    "sinon": "^6.0.0"
   }
```
Comment (abhayks1) [inline package.json]: Let's also freeze these npm versions.

**Review by abhayks1 — CHANGES_REQUESTED** [review]

**Review by kedarchandrayan — APPROVED** [review]
All the changes which requested by me earlier have been incorporated. The PR looks good for merge.

File: test/unit/lib/proof/storage_proof_test.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  return {
+    findPath: function (path, callback) {
+      callback(null, generateValidProof ? {"value": Buffer.from(proof.value, 'hex')} : undefined, "", utils.decodeParentNodes(proof.rlpParentNodes));
+    },
+    root: Buffer.from(proof.root, 'hex')
+  };
+}
+
+describe('should generate storage proof for  mapping type variable', function () {
+  let storageProofInstance, StorageProof, storageProof;
```
Comment (deepesh-kn) [inline test/unit/lib/proof/storage_proof_test.js]: variable name `StorageProof` and `storageProof` is creating confusion. We should name it better.

File: test/unit/lib/proof/account_proof_test.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      callback(null, generateValidProof ? utils.accountValue(proof.accountNode) : undefined, "", utils.decodeParentNodes(proof.rlpParentNodes));
+    },
+    root: Buffer.from(proof.root, 'hex')
+  };
+}
+
+describe('account proof for single node', function () {
+  let accountProofInstance, AccountProof,accountProof;
+  before(async () => {
+    AccountProof = require(rootPrefix + '/lib/proof/account_proof')
```
Comment (deepesh-kn) [inline test/unit/lib/proof/account_proof_test.js]: Any specific reason to have require here ? There are many such instances ex. line 48. (and also in other files.)

File: lib/proof/proof.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  /**
+   * Generate Storage proof for give storagePath in merkel patricia tree
+   * @param storagePath
+   * @param trie
+   * @return {Promise<proof>}
+   */
+  storageProof: function (storagePath, trie) {
+
+  let errorConf = {
+    internal_error_identifier: "l_p_accountProof_2",
```
Comment (deepesh-kn) [inline lib/proof/proof.js:60]: Not so critical but like to mention here.

General observation:  we follow a pattern to create internal_error_identifier. 
Lets take this example, file path is `lib/proof/proof.js`, function name is `storageProof`
So the internal_error_identifier will be in this case: `l_p_p_storageProof_1`

If we are still following pattern then we need to update `internal_error_identifier` at multiple locations.

**Review by deepesh-kn — COMMENTED** [review]
Added minor comments, other than variable name `StorageProof` and `storageProof` it looks good to me.

**Review by sunilkhedar — APPROVED** [review]
LGTM

**Pull request merged**